camunda-sdk-java/java-common/src/main/java/io/camunda/common/auth/SelfManagedAuthentication.java

```diff
@@ -81,65 +63,22 @@ public void resetToken(Product product) {
     tokens.remove(product);
   }
 
-  private String retrieveToken(Product product, JwtCredential jwtCredential) {
-    try(CloseableHttpClient client = HttpClients.createDefault()) {
-      HttpPost request = buildRequest(jwtCredential);
-      TokenResponse tokenResponse =
-          client.execute(
-              request,
-              response -> {
-                if (response.getCode() == HttpStatus.SC_OK) {
-                  return jsonMapper.fromJson(
-                      EntityUtils.toString(response.getEntity()), TokenResponse.class);
-                } else {
-                  throw new SdkException(
-                      "Error "
-                          + response.getCode()
-                          + " obtaining access token: "
-                          + EntityUtils.toString(response.getEntity()));
-                }
-              });
-      tokens.put(product, tokenResponse.getAccessToken());
-    } catch (Exception e) {
-      LOG.error("Authenticating for " + product + " failed due to " + e);
-      throw new SdkException("Unable to authenticate", e);
-    }
-    return tokens.get(product);
-  }
-
-  private HttpPost buildRequest(JwtCredential jwtCredential) {
-    HttpPost httpPost = new HttpPost(authUrl);
-    httpPost.addHeader("Content-Type", "application/x-www-form-urlencoded");
-
-    Map<String, String> parameters = new HashMap<>();
-    parameters.put("grant_type", "client_credentials");
-    parameters.put("client_id", jwtCredential.getClientId());
-    parameters.put("client_secret", jwtCredential.getClientSecret());
-
-    String form = parameters.entrySet()
-      .stream()
-      .map(e -> {
-        try {
-          return e.getKey() + "=" + URLEncoder.encode(e.getValue(), StandardCharsets.UTF_8.toString());
-        } catch (UnsupportedEncodingException ex) {
-          throw new RuntimeException(ex);
-        }
-      })
-      .collect(Collectors.joining("&"));
-
-    httpPost.setEntity(new StringEntity(form));
-
-    return  httpPost;
-  }
-
   @Override
   public Map.Entry<String, String> getTokenHeader(Product product) {
     String token;
     if (tokens.containsKey(product)) {
       token = tokens.get(product);
     } else {
-      JwtCredential jwtCredential = jwtConfig.getProduct(product);
-      token = retrieveToken(product, jwtCredential);
+      Identity identity = identityConfig.get(product).getIdentity();
+      String audience = jwtConfig.getProduct(product).getAudience();
+      Tokens identityTokens = identity.authentication().requestToken(audience);
+      try {
+        identity.authentication().verifyToken(identityTokens.getAccessToken());
+      } catch (TokenExpiredException exception) {
+        identityTokens = identity.authentication().renewToken(identityTokens.getRefreshToken());
+      }
```

Comment:
Do we expect that a token fetched just now is already expired?

Reply:
Thats a remote possibility but these verify logic here can be removed if Identity SDK has an ability to remove the token in their cache. Raised a [question](https://github.com/camunda-cloud/identity/issues/2496)/feature request on their board

Basically, whenever we received a 401 or expired token from some endpoints, we remove the token in memory in Spring SDK. However since we now rely on Identity SDK, we would also need to have the ability to evict token in Identity Cache too.

Right now, I'm attempting to remove the token in Spring SDK memory, however whenever we get a new token with Identity that was already expired, their logic would still retrieve the expired in cache. The current workaround would be get a new token thru referesh token, but as far as I know, the expired token always sits in their cache.

Reply:
I was just wondering as this code is only called after fetching a new token.